lib/astutils.cpp

```diff
@@ -1210,114 +1210,107 @@ SmallVector<ReferenceToken> followAllReferences(const Token* tok,
-                    if (n < 0) {
-                        refs_result.push_back({tok, std::move(errors)});
-                        return refs_result;
-                    }
-                    std::vector<const Token*> args = getArguments(tok->previous());
-                    if (n >= args.size()) {
-                        refs_result.push_back({tok, std::move(errors)});
-                        return refs_result;
+        } else if (tok->previous() && tok->previous()->function() && Token::Match(tok->previous(), "%name% (")) {
+            const Function *f = tok->previous()->function();
+            if (!Function::returnsReference(f)) {
+                break;
+            }
+            std::set<ReferenceToken, ReferenceTokenLess> result;
+            std::vector<const Token*> returns = Function::findReturns(f);
+            for (const Token* returnTok : returns) {
+                if (returnTok == tok)
+                    continue;
+                for (const ReferenceToken& rt :
+                     followAllReferences(returnTok, temporary, inconclusive, errors, depth - returns.size())) {
+                    const Variable* argvar = rt.token->variable();
+                    if (!argvar) {
+                        break;
                     }
-                    const Token* argTok = args[n];
-                    ErrorPath er = errors;
-                    er.emplace_back(returnTok, "Return reference.");
-                    er.emplace_back(tok->previous(), "Called function passing '" + argTok->expressionString() + "'.");
-                    auto refs =
-                        followAllReferences(argTok, temporary, inconclusive, std::move(er), depth - returns.size());
-                    result.insert(refs.cbegin(), refs.cend());
-                    if (!inconclusive && result.size() > 1) {
-                        refs_result.push_back({tok, std::move(errors)});
-                        return refs_result;
+                    if (argvar->isArgument() && (argvar->isReference() || argvar->isRValueReference())) {
+                        const int n = getArgumentPos(argvar, f);
+                        if (n < 0) {
+                            break;
+                        }
+                        std::vector<const Token*> args = getArguments(tok->previous());
+                        if (n >= args.size()) {
+                            break;
+                        }
+                        const Token* argTok = args[n];
+                        ErrorPath er = errors;
+                        er.emplace_back(returnTok, "Return reference.");
+                        er.emplace_back(tok->previous(), "Called function passing '" + argTok->expressionString() + "'.");
+                        auto refs =
+                            followAllReferences(argTok, temporary, inconclusive, std::move(er), depth - returns.size());
+                        if (!inconclusive && refs.size() > 1) {
```

Comment:
This behavior for this and the following lines has also changed. Again I missed the outer loop so `result` might have been set in an earlier iteration.

But if we hit this block more than once we bailed out when the prerequisites were not met even though we might already have entries in `result`. If that should/can never happen or is unintentional we also need to exit after we added something to `result`.

Reply:
Or we should have continued instead of returned.